src/storage/index/index_builder.cpp

```diff
@@ -50,6 +49,58 @@ IndexBuilder &IndexBuilder::SetKeySchema(const catalog::IndexSchema &key_schema)
   return *this;
 }
 
+IndexBuilder &IndexBuilder::SetSqlTableAndTransactionContext(
+    common::ManagedPointer<transaction::TransactionContext> txn, common::ManagedPointer<storage::SqlTable> sql_table,
+    const catalog::IndexSchema &key_schema) {
+  TERRIER_ASSERT((sql_table == nullptr && txn == nullptr) || (sql_table != nullptr && txn != nullptr),
+                 "sql_table / txn is null and txn / sql_table is not.");
+  txn_ = txn;
+  sql_table_ = sql_table;
+  key_schema_ = key_schema;
+  return *this;
+}
+
+bool IndexBuilder::Insert(common::ManagedPointer<storage::index::Index> index,
+                          storage::TupleSlot table_tuple_slot) const {
+  // Initialize index pr
+  const auto index_pr_initializer = index->GetProjectedRowInitializer();
+  const uint32_t index_pr_size = index_pr_initializer.ProjectedRowSize();
+  byte *index_pr_buffer = common::AllocationUtil::AllocateAligned(index_pr_size);
+  ProjectedRow *index_pr = index_pr_initializer.InitializeRow(index_pr_buffer);
+  // get all index col oid and Initialize table pr
+  const auto &indexed_attributes = key_schema_.GetIndexedColOids();
+  const auto table_pr_initializer = sql_table_->InitializerForProjectedRow(indexed_attributes);
```

Comment:
This is a computationally expensive operation to perform on every insert. It should be stashed at least one abstraction level up (outside of the loop). That's true for basically all of the metadata that doesn't change for every insert. We should just be doing the PR population and index insert in a tight loop.

Reply:
moved outside of the loop